### PR TITLE
Remove RemoteCacheManager per URI logic

### PR DESCRIPTION
**Pull request by ryanemerson**

### CODE REVIEW

File: src/main/java/io/gingersnapproject/cdc/ManagedEngine.java [hunk 1]
```diff
@@ -38,25 +32,17 @@ public class ManagedEngine {
...
-            return cacheService;
-         }
-      }
-      throw new IllegalArgumentException("Unsupported URI received: " + uri + " ensure service is running if correct!");
-   }
+   @Inject NotificationManager eventing;
 
    public void start(@Observes StartupEvent ignore) {
       log.info("Starting service");
+      cacheService.start();
```
Comment (ryanemerson) [inline src/main/java/io/gingersnapproject/cdc/ManagedEngine.java:42]: Is there a better way we can handle this? It doesn't seem great controlling a bean's lifecycle from another bean.

Comment (wburns) [reply src/main/java/io/gingersnapproject/cdc/ManagedEngine.java:42]: You can add the same events to the CacheService impls instead. Then they can control their own lifecycle. Then iiuc  quarkus will order based on dependency. So in this case the CacheService is a dependency of the Engine so they should be stopped in appropriate order.

File: src/main/java/io/gingersnapproject/health/CacheBackendReadiness.java [hunk 2]
```diff
@@ -4,32 +4,37 @@
...
 @Readiness
 @ApplicationScoped
 public class CacheBackendReadiness extends AbstractHealthChecker<URI> {
    private static final Logger log = LoggerFactory.getLogger(CacheBackendReadiness.class);
    private static final String BACKEND_HEALTH = "Backend Health";
 
+   @Inject
+   Configuration config;
+
    void backendStarted(@Observes Events.BackendStartedEvent ev) {
```
Comment (wburns) [inline src/main/java/io/gingersnapproject/health/CacheBackendReadiness.java:25]: It looks like this eventing can be redone as well now as we don't have multiple uris.

File: src/main/java/io/gingersnapproject/cdc/remote/RemoteOffsetStore.java [hunk 3]
```diff
@@ -64,13 +62,10 @@ public void configure(WorkerConfig workerConfig) {
...
-         if (cacheService.supportsURI(uri)) {
-            offsetBackend = cacheService.offsetBackendForURI(uri);
-            break;
-         }
-      }
+
+      CacheService cacheService = ArcUtil.instance(CacheService.class);
+      offsetBackend = cacheService.offsetBackendForURI();
+      // TODO update
       if (offsetBackend == null) {
```
Comment (wburns) [inline src/main/java/io/gingersnapproject/cdc/remote/RemoteOffsetStore.java:69]: I don't believe this should be possible anymore right?

File: src/main/java/io/gingersnapproject/cdc/remote/RemoteSchemaHistory.java [hunk 4]
```diff
@@ -56,13 +55,8 @@ public void configure(Configuration config, HistoryRecordComparator comparator,
...
-      for (InstanceHandle<CacheService> instanceHandle : Arc.container().listAll(CacheService.class)) {
-         CacheService cacheService = instanceHandle.get();
-         if (cacheService.supportsURI(uri)) {
-            schemaBackend = cacheService.schemaBackendForURI(uri);
-            break;
-         }
-      }
+      CacheService cacheService = ArcUtil.instance(CacheService.class);
+      schemaBackend = cacheService.schemaBackendForURI();
       if (schemaBackend == null) {
```
Comment (wburns) [inline src/main/java/io/gingersnapproject/cdc/remote/RemoteSchemaHistory.java:60]: I don't believe this should be possible anymore right?

File: src/main/java/io/gingersnapproject/cdc/EngineWrapper.java [hunk 5]
```diff
@@ -118,19 +118,19 @@ public void taskStopped() {
...
       engine = null;
-      cacheService.stop(config.cache().uri());
+      cacheService.stop();
    }
 
    public void notifyError(Throwable t) {
       eventing.connectorFailed(name, t);
    }
 
    public void shutdownCacheService() {
```
Comment (wburns) [inline src/main/java/io/gingersnapproject/cdc/EngineWrapper.java:128]: I don't think this stuff makes sense anymore either if we have a single configured cache service that we are bound to. If anything it should be controlled solely by the config watcher / rest endpoints to change the URI right?

**Review by wburns — COMMENTED** [review]
I think we need to take this a bit further with the changes.

File: src/main/java/io/gingersnapproject/cdc/ManagedEngine.java (hunk 1 above)
Comment (ryanemerson) [reply src/main/java/io/gingersnapproject/cdc/ManagedEngine.java:42]: I did try that originally, but on some startups the CacheService wasn't initated before `ManagedEngine#startStopEngine` was called.

File: src/main/java/io/gingersnapproject/health/CacheBackendReadiness.java (hunk 2 above)
Comment (ryanemerson) [reply src/main/java/io/gingersnapproject/health/CacheBackendReadiness.java:25]: We don't have multiple URIs, but passing the global one means that the health is still represented as:

```json
        {
            "name": "Backend Health",
            "status": "DOWN",
            "data": {
                "hotrod://127.0.0.1:11223": false
            }
        }
```

File: src/main/java/io/gingersnapproject/cdc/EngineWrapper.java (hunk 5 above)
Comment (ryanemerson) [reply src/main/java/io/gingersnapproject/cdc/EngineWrapper.java:128]: I don't think we need to support changing the URI at runtime at all tbh. If the ServiceBinding is updated then so is the pod.

**Comment by ryanemerson** [comment]
@wburns I'll take a closer look tomorrow to see if I can make it so the CacheService bean's lifecycle is fully independent.

File: src/main/java/io/gingersnapproject/cdc/ManagedEngine.java (hunk 1 above)
Comment (wburns) [reply src/main/java/io/gingersnapproject/cdc/ManagedEngine.java:42]: Weird.. that sounds like a bug in Quarkus. I would expect dependencies to have start method called in dependent order. But maybe they don't :(

Comment (jabolina) [reply src/main/java/io/gingersnapproject/cdc/ManagedEngine.java:42]: Another alternative would be creating a method with the `@Produces`, starting the service, and returning it.

Comment (ryanemerson) [reply src/main/java/io/gingersnapproject/cdc/ManagedEngine.java]: I have got it working now, I just needed to make the HotRodService `@ApplicationScoped` and pass it as a parameter of the ManagedEngine#start

File: src/main/java/io/gingersnapproject/cdc/EngineWrapper.java (hunk 5 above)
Comment (wburns) [reply src/main/java/io/gingersnapproject/cdc/EngineWrapper.java]: Sounds good to me.

File: src/main/java/io/gingersnapproject/cdc/cache/hotrod/HotRodService.java [hunk 6]
```diff
@@ -30,48 +27,60 @@
...
 
+   public void shutdown(@Observes ShutdownEvent e) {
+      if (rcm != null) {
+         rcm.stopAsync().whenComplete((ignore, t) -> {
+            if (t != null) {
+               eventing.backendFailedEvent(t);
+            } else {
+               eventing.backendStoppedEvent();
+            }
+         });
```
Comment (wburns) [inline src/main/java/io/gingersnapproject/cdc/cache/hotrod/HotRodService.java:56]: I am not 100% on the lifecycle of quarkus, but maybe set rcm to null here as well.

**Review by wburns — APPROVED** [review]
Just a very minor comment, otherwise looks good to me.

**Comment by ryanemerson** [comment]
Thanks for reviewing @jabolina @wburns

**Pull request merged**